From dd09697fa178321df7345a3232da7d01890cb790 Mon Sep 17 00:00:00 2001
From: "guoyin.chen" <guoyin.chen@freescale.com>
Date: Wed, 6 Feb 2013 18:23:21 +0800
Subject: [PATCH 418/498] ENGR00243668 testCameraToSurfaceTextureMetadata
 Failed

Preview stream will not be reallocated if resolution not changed
in the CTS. CameraFrame's reference will not be re-inited in this
case. There will be two buffers hold in NativeWindow which need to
take care their buffer refCount

Signed-off-by: guoyin.chen <guoyin.chen@freescale.com>
---
 mx6/libcamera2/CameraUtil.cpp    | 5 +++++
 mx6/libcamera2/CameraUtil.h      | 1 +
 mx6/libcamera2/PreviewStream.cpp | 7 +++++--
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/mx6/libcamera2/CameraUtil.cpp b/mx6/libcamera2/CameraUtil.cpp
index d0e0c60..d72dd2c 100755
--- a/mx6/libcamera2/CameraUtil.cpp
+++ b/mx6/libcamera2/CameraUtil.cpp
@@ -170,6 +170,11 @@ void CameraFrame::release()
     }
 }
 
+int CameraFrame::getRefCount()
+{
+    return mRefCount;
+}
+
 void CameraFrame::setObserver(CameraFrameObserver *observer)
 {
     mObserver = observer;
diff --git a/mx6/libcamera2/CameraUtil.h b/mx6/libcamera2/CameraUtil.h
index 6ff35fb..c8c4635 100755
--- a/mx6/libcamera2/CameraUtil.h
+++ b/mx6/libcamera2/CameraUtil.h
@@ -208,6 +208,7 @@ public:
     void removeState(CAMERA_BUFS_STATE state);
     void release();
     void addReference();
+    int getRefCount();
     void setObserver(CameraFrameObserver *observer);
     void reset();
 
diff --git a/mx6/libcamera2/PreviewStream.cpp b/mx6/libcamera2/PreviewStream.cpp
index e46eda0..77b1a78 100755
--- a/mx6/libcamera2/PreviewStream.cpp
+++ b/mx6/libcamera2/PreviewStream.cpp
@@ -82,8 +82,11 @@ int PreviewStream::allocateBuffers(int width, int height,
     for (int i = 0; i < mTotalBuffers; i++) {
         int state = mCameraBuffer[i].getState();
         if (state == CameraFrame::BUFS_IN_SERVICE) {
-            //the frame held in service.
-            mCameraBuffer[i].addReference();
+            // The frame held in service.
+            // Make sure we dont add one more reference
+            // count for it
+            if(!mCameraBuffer[i].getRefCount())
+                mCameraBuffer[i].addReference();
         }
     }
 
-- 
1.8.0

